queue/example_connection_pool_test.go

```diff
@@ -0,0 +1,148 @@
+package queue_test
+
+import (
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/tarantool/go-tarantool"
+	"github.com/tarantool/go-tarantool/connection_pool"
+	"github.com/tarantool/go-tarantool/queue"
+)
+
+type QueueConnectionHandler struct {
+	name string
+	cfg  queue.Cfg
+
+	uuid          uuid.UUID
+	registered    bool
+	err           error
+	registerMutex sync.Mutex
+	done          chan struct{}
+}
+
+var _ connection_pool.ConnectionHandler = &QueueConnectionHandler{}
+
+func NewQueueConnectionHandler(name string, cfg queue.Cfg) *QueueConnectionHandler {
+	return &QueueConnectionHandler{
+		name: name,
+		cfg:  cfg,
+		done: make(chan struct{}),
+	}
+}
+
+func (h *QueueConnectionHandler) Discovered(conn *tarantool.Connection,
+	role connection_pool.Role) error {
+	// The queue only works with a master instance.
+	if role != connection_pool.MasterRole {
+		return nil
+	}
+
+	// We need to register a shared session only once.
+	h.registerMutex.Lock()
+
+	if h.err != nil {
+		h.registerMutex.Unlock()
+		return h.err
+	}
+
+	q := queue.New(conn, h.name)
+	if !h.registered {
+		opts := queue.CfgOpts{InReplicaset: true, Ttr: 60 * time.Second}
+		if h.err = q.Cfg(opts); h.err == nil {
+			if h.uuid, h.err = q.Identify(nil); h.err == nil {
+				h.err = q.Create(h.cfg)
+			}
+		}
+		h.registered = h.err == nil
+		close(h.done)
+		h.registerMutex.Unlock()
+		return h.err
+	} else {
+		// We already registered. We don't need the lock any more.
+		h.registerMutex.Unlock()
+		if _, err := q.Identify(&h.uuid); err != nil {
+			return err
+		}
+		if err := q.Create(h.cfg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *QueueConnectionHandler) Deactivated(conn *tarantool.Connection,
+	role connection_pool.Role) error {
+	return nil
+}
+
+// Example demonstrates how to use the queue package with the connection_pool
+// package. First of all, you need to create a connection handler for the
+// a ConnectionPool object to process new connections from RW-instances.
+//
+// You need to registry a shared session UUID at a first master connection.
+// It needs to be used to re-identify as the shared session on new
+// RW-instances. See QueueConnectionHandler.Discovered() implementation.
+//
+// After that, you need to create a ConnectorAdapter object with RW mode for
+// the ConnectionPool to send requests into RW-instances. This adapter can
+// be used to create a ready-to-work queue object.
+func Example_connectionPool() {
```

Comment:
Why it is an example and not some API method that user can call in one line? May user want to change some parts of the example?

Reply:
ConnectionPool has much more use cases (master-master as example, just instances without a replicaset). I think it is impossible to provide a universal API here.

Reply:
Does queue support master-master?

Reply:
It looks like it doesn't, at least I haven't seen any collision protection.